input/listen.go

```diff
@@ -8,126 +8,161 @@ import (
 	"github.com/jpillora/backoff"
 )
 
-func listen(addr string, handler Handler) error {
+type Listener struct {
+	workerPool
+	tcpList *net.TCPListener
+	udpConn *net.UDPConn
+	handler Handler
+}
+
+func NewListener(handler Handler) *Listener {
+	return &Listener{
+		handler: handler,
+	}
+}
+
+func (l *Listener) listen(addr string) error {
+	l.startStoppable()
+
 	// listeners are set up outside of accept* here so they can interrupt startup
-	listener, err := listenTcp(addr)
+	err := l.listenTcp(addr)
 	if err != nil {
 		return err
 	}
 
-	udp_conn, err := listenUdp(addr)
+	err = l.listenUdp(addr)
 	if err != nil {
 		return err
 	}
 
-	go acceptTcp(addr, listener, handler)
-	go acceptUdp(addr, udp_conn, handler)
+	l.wg.Add(2)
+	go l.accept(addr, "tcp", l.acceptTcp, l.listenTcp, l.tcpList)
+	go l.accept(addr, "udp", l.consumeUdp, l.listenUdp, l.udpConn)
 
 	return nil
 }
 
-func listenTcp(addr string) (*net.TCPListener, error) {
-	laddr, err := net.ResolveTCPAddr("tcp", addr)
-	if err != nil {
-		return nil, err
-	}
-	return net.ListenTCP("tcp", laddr)
-}
+func (l *Listener) accept(addr, proto string, consume func(string), reconnect func(string) error, listener Closable) {
```

Comment:
why is this function called accept? something like `run` seems more appropriate and less confusing.